libraries/ms-drivers/test/test_spv1020_mppt.c

```diff
@@ -0,0 +1,136 @@
+#include "log.h"
+#include "spv1020_mppt.h"
+#include "spv1020_mppt_defs.h"
+#include "test_helpers.h"
+#include "unity.h"
+
+#define TEST_SPI_PORT SPI_PORT_2
+
+#define TEST_BAUDRATE 60000
+#define TEST_MOSI_PIN \
+  { .port = GPIO_PORT_B, 15 }
+#define TEST_MISO_PIN \
+  { .port = GPIO_PORT_B, 14 }
+#define TEST_SCLK_PIN \
+  { .port = GPIO_PORT_B, 13 }
+#define TEST_CS_PIN \
+  { .port = GPIO_PORT_B, 12 }
+
+void setup_test() {
+  gpio_init();
+  SpiSettings spi_settings = {
+    .baudrate = 60000,
+    .mode = SPI_MODE_0,
+    .mosi = TEST_MOSI_PIN,
+    .miso = TEST_MISO_PIN,
+    .sclk = TEST_SCLK_PIN,
+    .cs = TEST_CS_PIN,
+  };
+  spi_init(TEST_SPI_PORT, &spi_settings);
+}
+void teardown_test() {}
+
+// Test that we can successfully send the SHUT and Turn ON commands.
+void test_shut_and_turn_on(void) {
+  TEST_ASSERT_OK(spv1020_shut(TEST_SPI_PORT));
+  TEST_ASSERT_OK(spv1020_turn_on(TEST_SPI_PORT));
+  TEST_ASSERT_OK(spv1020_shut(TEST_SPI_PORT));
+  TEST_ASSERT_OK(spv1020_shut(TEST_SPI_PORT));  // 2 in a row, there should be a warning on x86
+  TEST_ASSERT_OK(spv1020_turn_on(TEST_SPI_PORT));
+  TEST_ASSERT_OK(spv1020_turn_on(TEST_SPI_PORT));  // unnecessary, there should be a warning on x86
+}
+
+// Test that we can successfully read a current and it's in the stated range.
+void test_read_current(void) {
+  uint16_t current = 0xFFFF;  // over 10 bits
+  TEST_ASSERT_OK(spv1020_read_current(TEST_SPI_PORT, &current));
+
+  // make sure the 6 highest bits are low, so it's a 10-bit value
+  TEST_ASSERT_BITS_LOW(0b1111110000000000, current);
+}
+
+// Test that we can successfully read a voltage and it's in the stated range.
+void test_read_voltage_in(void) {
+  uint16_t vin = 0xFFFF;  // over 10 bits
+  TEST_ASSERT_OK(spv1020_read_voltage_in(TEST_SPI_PORT, &vin));
+
+  // make sure the 6 highest bits are low, so it's a 10-bit value
+  TEST_ASSERT_BITS_LOW(0b1111110000000000, vin);
+}
+
+// Test that we can successfully read PWM and it's in the stated range.
+void test_read_pwm(void) {
+  uint16_t pwm = 0xFFFF;  // over 9 bits
+  TEST_ASSERT_OK(spv1020_read_pwm(TEST_SPI_PORT, &pwm));
+
+  // make sure the 6 highest bits are low, so it's a 9-bit value
+  TEST_ASSERT_BITS_LOW(0b1111111000000000, pwm);
```

Comment:
could add an assertion that pwm is less than expected number of steps rather than just a 9 bit number, depends on if you add the conversion to actual pwm from the 9 bits.